pam_okta/pam_okta.8
.\"
.\" Copyright (c) 2026 The University of Queensland
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: July 16 2013 $
.Dt PAM_OKTA 8
.Os
.Sh NAME
.Nm pam_okta
.Nd Module for using Okta for authentication
.Sh SYNOPSIS
.Nm
.Op Ar ...
.Sh DESCRIPTION
This authentication module authenticates users against the Okta API.
.Pp
Rather than the
.Nm
module talking directly to the Okta API, it connects to a
.Xr pam_oktad 8
daemon that communicates on the modules behalf.
.Sh OPTIONS
.Bl -tag -width indent
.It Cm mode Ns = Ns Ar authmode
Specifies which Okta authentication flow to use.
The supported modes are:
.Bl -inset
.It Cm mfa-oob
Authenticate the user with their password, and depending on the Okta
service configuration, an out-of-band (OOB) push factor.
It follows the Direct Authentication using the Okta Verify Push (MFA)
flow.
This is the default mode.
.It Cm oob
Authenticate the user with an out-of-band (OOB) push factor.
It follows the Direct Authentication using the Okta Verify Push
(primary factor) flow.
This mode does not verify the users password.
If password authentication is required, a preceding module in the stack
must check it, or use the
.Cm mfa-oob
mode.
.It Cm mfa-otp
Authenticate the user with their password, and depending on the Okta
service configuration, a one time password (OTP).
It follows the Direct Authentication using the OTP (MFA) flow.
.It Cm otp
Authenticate the user with a one time password (OTP).
It follows the Direct Authentication using the OTP (primary factor) flow.
This mode does not verify the users password.
If password authentication is required, a preceding module in the stack
must check it, or use the
.Cm mfa-otp
mode.
.It Cm mfa-oob-otp
Authenticate the user with their password, and depending on the Okta
service configuration, an out-of-band (OOB) push factor or one time
password (OTP).
.It Cm oob-otp
Authenticate the user with an out-of-band (OOB) push factor or one time
password (OTP).
This mode does not verify the users password.
If password authentication is required, a preceding module in the stack
must check it, or use the
.Cm mfa-oob-otp
mode.
.It Cm device
Use the Device Authentication flow to authenticate the user.
This mode does not verify the users password.
If password authentication is required, a preceding module in the stack
must check it.
.El
.It Cm allow_decline
Allow the user to decline out-of-band (OOB) authentication.
.It Cm socket Ns = Ns Ar sockname
Specify the path of the Unix Domain Socket that the
.Xr pam_oktad 8
daemon is listening on for authentication requests from this module.
The default
.Ar sockname
is
.Va /var/run/pam_okta/sock .
.It Cm debug
Enable extra debug logging.
.It Cm no_warn
Disable warnings when possible.
.It Cm try_first_pass
The module will try and use an existing password before prompting the user.
This is only used when the
.Cm mode
is
.Cm mfa-oob .
.It Cm use_first_pass
The module will try and use an existing password.
If no password is available or is not appropriate, authentication will fail.
This is only used when the
.Cm mode
is
.Cm mfa-oob .
.It Cm sshd Ns = Ns Ar servicename
.Nm
sends the
.Ev SSH_CONNECTION
environment variable to
.Xr pam_oktad 8
if the PAM service attempting authentication is
.Va sshd .
The
.Cm sshd
option enables this behaviour for an alternative PAM service name.
The default value is
.Va sshd .
.El
.Sh MODULE TYPES PROVIDED
.Nm
only provides
auth.
.Sh EXAMPLES
An example usage for /etc/pam.d/login would be:
.Bd -literal -offset indent
# Authenticate the user
auth       required   pam_okta.so mode=mfa-oob
# Ensure user's account and password are still active
account    required   pam_unix.so
.Ed
.Sh SEE ALSO
.Xr login.defs 5 ,
.Xr pam.conf 5 ,
.Xr pam.d 5 ,
.Xr pam_oktad 8 ,
.Xr pam 8
.\" .Sh AUTHOR
.\" pam_unix was written by various people.


pam_okta/pam_okta.c
/* */

/*
 * Copyright (c) 2026 The University of Queensland
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/socket.h>
#include <sys/un.h>
#include <sys/uio.h>

#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <syslog.h>
#include <errno.h>

#define PAM_SM_AUTH
#define PAM_SM_ACCOUNT
#define PAM_SM_SESSION
#define PAM_SM_PASSWORD

#include <security/pam_modules.h>
#include <security/pam_ext.h>

#include "protocol.h"

#define nitems(_a) (sizeof((_a)) / sizeof((_a)[0]))

struct state {
	pam_handle_t	*pamh;
	unsigned int	 pamflags;
#define CFG_F_DEBUG		(1 << 0)
#define CFG_F_NO_WARN		(1 << 1)
	unsigned int	 first_pass;
#define CFG_FIRST_PASS_UNSET	0
#define CFG_FIRST_PASS_USE	1
#define CFG_FIRST_PASS_TRY	2
	unsigned int	 mode;
	unsigned int	 flags;

	const char	*sockname;
	const char	*sshd; /* which service provides SSH_CONNECTION env */

	const char	*service;
	size_t		 servicelen;
	const char	*user;
	size_t		 userlen;
	const char	*pass;
	size_t		 passlen;
	const char	*rhost;
	size_t		 rhostlen;
	const char	*sshenv;
	size_t		 sshenvlen;
};

static const char sockopt[] = "socket=";
#define SOCKOPTLEN (sizeof(sockopt) - 1)
static const char sshdopt[] = "sshd=";
#define SSHDOPTLEN (sizeof(sshdopt) - 1)

static int
parse_args(struct state *st, int argc, const char **argv)
{
	int i;

	for (i = 0; i < argc; i++) {
		if (strcmp(argv[i], "debug") == 0) {
			st->pamflags |= CFG_F_DEBUG;
		} else if (strcmp(argv[i], "no_warn") == 0) {
			st->pamflags |= CFG_F_NO_WARN;
		} else if (strcmp(argv[i], "use_first_pass") == 0) {
			st->first_pass = CFG_FIRST_PASS_USE;
		} else if (strcmp(argv[i], "try_first_pass") == 0) {
			st->first_pass = CFG_FIRST_PASS_TRY;
		} else if (strcmp(argv[i], "mode=mfa-oob") == 0) {
			st->mode = OKTA_MODE_MFA_OOB;
		} else if (strcmp(argv[i], "mode=oob") == 0) {
			st->mode = OKTA_MODE_OOB;
		} else if (strcmp(argv[i], "mode=device") == 0) {
			st->mode = OKTA_MODE_DEVICE_AUTH;
		} else if (strcmp(argv[i], "mode=mfa-otp") == 0) {
			st->mode = OKTA_MODE_MFA_OTP;
		} else if (strcmp(argv[i], "mode=otp") == 0) {
			st->mode = OKTA_MODE_OTP;
		} else if (strcmp(argv[i], "mode=mfa-oob-otp") == 0) {
			st->mode = OKTA_MODE_MFA_OOB_OTP;
		} else if (strcmp(argv[i], "mode=oob-otp") == 0) {
			st->mode = OKTA_MODE_OOB_OTP;
		} else if (strcmp(argv[i], "allow_decline") == 0) {
			st->flags |= OKTA_F_ALLOW_DECLINE;
		} else if (strncmp(argv[i], sockopt, SOCKOPTLEN) == 0) {
			st->sockname = argv[i] + SOCKOPTLEN;
			if (st->sockname[0] != '/') {
				pam_syslog(st->pamh, LOG_ERR,
				    "socket name must be an absolute path");
				return (-1);
			}
		} else if (strncmp(argv[i], sshdopt, SSHDOPTLEN) == 0) {
			st->sshd = argv[i] + SSHDOPTLEN;
		} else {
			pam_syslog(st->pamh, LOG_ERR,
			    "unexpected argument %s", argv[i]);
			return (-1);
		}
	}

	return (0);
}

static int
okta_authn_password(pam_handle_t *pamh, struct state *st)
{
	char *pass;
	const void *item;
	int rv;

	switch (st->first_pass) {
	case CFG_FIRST_PASS_UNSET:
		rv = pam_prompt(pamh, PAM_PROMPT_ECHO_OFF,
		    &pass, "Password: ");
		if (rv != PAM_SUCCESS) {
			pam_syslog(pamh, LOG_ERR,
			    "password prompt: %s", pam_strerror(pamh, rv));
			return (rv);
		}
		if (pass == NULL) {
			rv = PAM_CONV_ERR;
			pam_syslog(pamh, LOG_ERR,
			    "password prompt: %s", pam_strerror(pamh, rv));
			return (rv);
		}

		rv = pam_set_item(pamh, PAM_AUTHTOK, pass);
		explicit_bzero(pass, strlen(pass));
		free(pass);
		if (rv != PAM_SUCCESS) {
			pam_syslog(pamh, LOG_ERR,
			    "set authtok: %s", pam_strerror(pamh, rv));
			return (rv);
		}

		/* FALLTHROUGH */
	case CFG_FIRST_PASS_USE:
		rv = pam_get_item(pamh, PAM_AUTHTOK, &item);
		if (rv != PAM_SUCCESS) {
			pam_syslog(pamh, LOG_ERR,
			    "get authtok: %s", pam_strerror(pamh, rv));
			return (rv);
		}
		st->pass = item;
		break;

	case CFG_FIRST_PASS_TRY:
		rv = pam_get_authtok(pamh, PAM_AUTHTOK, &st->pass, NULL);
		if (rv != PAM_SUCCESS) {
			pam_syslog(pamh, LOG_ERR,
			    "get authtok: %s", pam_strerror(pamh, rv));
			return (rv);
		}
		break;
	default:
		abort();
		/* NOTREACHED */
	}

	if (st->pass == NULL) {
		rv = PAM_BAD_ITEM;
		pam_syslog(pamh, LOG_ERR,
		    "get authtok: %s", pam_strerror(pamh, rv));
		return (rv);
	}

	st->passlen = strlen(st->pass) + 1;

	return (PAM_SUCCESS);
}

static int
okta_authn_prepare(struct state *st)
{
	pam_handle_t *pamh = st->pamh;
	const void *item;
	int rv;

	rv = pam_get_item(pamh, PAM_SERVICE, &item);
	if (rv != PAM_SUCCESS) {
		pam_syslog(pamh, LOG_ERR,
		    "PAM_SERVICE: %s", pam_strerror(pamh, rv));
		return (rv);
	}
	if (item == NULL) {
		pam_syslog(pamh, LOG_ERR, "PAM_SERVICE is not set");
		return (PAM_SERVICE_ERR);
	}
	st->service = item;
	st->servicelen = strlen(st->service) + 1;

	rv = pam_get_user(pamh, &st->user, NULL);
	if (rv != PAM_SUCCESS) {
		pam_syslog(pamh, LOG_ERR,
		    "pam_get_user: %s", pam_strerror(pamh, rv));
		return (rv);
	}
	st->userlen = strlen(st->user) + 1;

	/* Okta Verify Push (MFA) starts with the username and password */
	switch (st->mode) {
	case OKTA_MODE_MFA_OOB:
	case OKTA_MODE_MFA_OTP:
	case OKTA_MODE_MFA_OOB_OTP:
		rv = okta_authn_password(pamh, st);
		if (rv != PAM_SUCCESS)
			return (rv);
		break;
	}

	rv = pam_get_item(pamh, PAM_RHOST, &item);
	if (rv != PAM_SUCCESS) {
		pam_syslog(pamh, LOG_ERR,
		    "PAM_RHOST: %s", pam_strerror(pamh, rv));
		return (rv);
	}
	if (item != NULL) {
		st->rhost = item;
		st->rhostlen = strlen(st->rhost) + 1;
	}

	if (strcmp(st->sshd, st->service) == 0) {
		st->sshenv = pam_getenv(pamh, "SSH_CONNECTION");
		if (st->sshenv != NULL)
			st->sshenvlen = strlen(st->sshenv) + 1;
	}

	return (PAM_SUCCESS);
}

static int
okta_connect(struct state *st)
{
	struct sockaddr_un sun = {
		.sun_family = AF_UNIX,
	};
	size_t socknamelen;
	int s;

	/* i miss strlcpy */
	socknamelen = strlen(st->sockname);
	if (socknamelen >= sizeof(sun.sun_path)) {
		pam_syslog(st->pamh, LOG_ERR,
		    "pam_oktad socket name is too long");
		return (-1);
	}

	s = socket(AF_UNIX, SOCK_SEQPACKET, 0);
	if (s == -1) {
		pam_syslog(st->pamh, LOG_ERR,
		    "socket: %s", strerror(errno));
		return (-1);
	}

	memcpy(sun.sun_path, st->sockname, socknamelen);
	sun.sun_path[socknamelen] = '\0';

	if (connect(s, (struct sockaddr *)&sun, sizeof(sun)) == -1) {
		pam_syslog(st->pamh, LOG_ERR,
		    "socket %s connect: %s", st->sockname, strerror(errno));
		close(s);
		return (-1);
	}

	return (s);
}

static int
okta_authn(struct state *st, int s)
{
	struct ctl_authn_req r = {
		.hdr = { .type = CTL_T_AUTHN_REQ, .hdrlen = sizeof(r) },
		.version = { .major = 0, .minor = 0 },
		.mode = st->mode,
		.flags = st->flags,

		.fields = {
			[CTL_AUTHN_REQ_SERVICE] = st->servicelen,
			[CTL_AUTHN_REQ_USERNAME] = st->userlen,
			[CTL_AUTHN_REQ_PASSWORD] = st->passlen,
			[CTL_AUTHN_REQ_RHOST] = st->rhostlen,
			[CTL_AUTHN_REQ_SSHENV] = st->sshenvlen,
		},
	};
	struct iovec iov[] = {
		{ &r, sizeof(r) },
		{ (void *)st->service, st->servicelen },
		{ (void *)st->user, st->userlen },
		{ (void *)st->pass, st->passlen },
		{ (void *)st->rhost, st->rhostlen },
		{ (void *)st->sshenv, st->sshenvlen },
	};
	ssize_t rv;

	rv = writev(s, iov, nitems(iov));
	if (rv == -1) {
		pam_syslog(st->pamh, LOG_ERR, "socket %s write authn: %s",
		    st->sockname, strerror(errno));
		return (PAM_AUTHINFO_UNAVAIL);
	}

	return (PAM_SUCCESS);
}

static int
okta_respond(pam_handle_t *pamh, struct state *st, int s,
    char *str, size_t len)
{
	struct ctl_authn_res res = {
		.hdr = { .type = CTL_T_AUTHN_RES, .hdrlen = sizeof(res) },
		.code = OKTA_CODE_PROMPT,
		.msglen = len,
	};
	struct iovec iov[2] = {
		{ &res, sizeof(res) },
		{ str, len },
	};
	ssize_t rv;

	rv = writev(s, iov, nitems(iov));
	if (rv == -1) {
		pam_syslog(st->pamh, LOG_ERR, "socket %s write res: %s",
		    st->sockname, strerror(errno));
		return (PAM_SERVICE_ERR);
	}

	return (PAM_INCOMPLETE); /* loop */
}

static int
okta_prompt(pam_handle_t *pamh, struct state *st, int s, const char *prompt)
{
	char *p = NULL;
	int rv;
	size_t len;

	rv = pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &p, "%s", prompt);
	if (rv != PAM_SUCCESS)
		return (rv);
	if (p == NULL)
		return (PAM_SERVICE_ERR);

	len = strlen(p) + 1;

	rv = okta_respond(pamh, st, s, p, len);

	explicit_bzero(p, len);
	free(p);

	return (rv);
}

static int
okta_loop(pam_handle_t *pamh, struct state *st, int s)
{
	char buf[65536];
	char *str = NULL;
	struct ctl_authn_res *res = (struct ctl_authn_res *)buf;
	ssize_t rv;
	size_t len;

	rv = read(s, buf, sizeof(buf));
	switch (rv) {
	case -1:
		pam_syslog(pamh, LOG_ERR,
		    "socket %s read: %s", st->sockname, strerror(errno));
		return (PAM_SERVICE_ERR);
	case 0:
		pam_syslog(pamh, LOG_DEBUG,
		    "socket %s disconnected", st->sockname);
		return (PAM_SERVICE_ERR);
	default:
		break;
	}

	len = rv;
	if (len < sizeof(res->hdr)) {
		pam_syslog(pamh, LOG_ERR,
		    "socket %s short hdr read (%zu < %zu bytes)", st->sockname,
		    len, sizeof(res->hdr));
		return (PAM_SERVICE_ERR);
	}
	if (res->hdr.type != CTL_T_AUTHN_RES) {
		pam_syslog(pamh, LOG_ERR,
		    "socket %s unexpected res type %u", st->sockname,
		    res->hdr.type);
		return (PAM_SERVICE_ERR);
	}
	if (res->hdr.hdrlen < sizeof(*res)) {
		pam_syslog(pamh, LOG_ERR,
		    "socket %s short res hdrlen", st->sockname);
		return (PAM_SERVICE_ERR);
	}
	if (len < sizeof(*res)) {
		pam_syslog(pamh, LOG_ERR,
		    "socket %s short res hdr", st->sockname);
		return (PAM_SERVICE_ERR);
	}

	len -= res->hdr.hdrlen;

	if (len < res->msglen) {
		pam_syslog(pamh, LOG_ERR,
		    "socket %s short res msg (%u < %zu)",
		    st->sockname, res->msglen, len);
		return (PAM_SERVICE_ERR);
	}
	len = res->msglen;
	if (len > 0) {
		str = buf + res->hdr.hdrlen;
		len--; /* remove nul byte accounting */
		if (str[len] != '\0') {
			pam_syslog(pamh, LOG_ERR,
			    "socket %s bad msg string", st->sockname);
			return (PAM_SERVICE_ERR);
		}

		pam_syslog(pamh, LOG_DEBUG,
		    "socket %s msg %s", st->sockname, str);
	}

	switch (res->code) {
	case OKTA_CODE_PROMPT:
		if (str == NULL || len < 1) {
			pam_syslog(pamh, LOG_ERR,
			    "socket %s prompt without msg", st->sockname);
			return (PAM_SERVICE_ERR);
		}
		break;
	case OKTA_CODE_SUCCESS:
		pam_syslog(pamh, LOG_DEBUG,
		    "socket %s success for user %s",
		    st->sockname, st->user);
		return (PAM_SUCCESS);
	case OKTA_CODE_DECLINE:
		pam_syslog(pamh, LOG_DEBUG,
		    "socket %s declined user %s",
		    st->sockname, st->user);
		return (PAM_USER_UNKNOWN);
	case OKTA_CODE_FAILURE:
		pam_syslog(pamh, LOG_DEBUG,
		    "socket %s auth failure for user %s",
		    st->sockname, st->user);
		return (PAM_AUTH_ERR);
	default:
		abort();
	}

	return (okta_prompt(pamh, st, s, str));
}

PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	struct state _st = { /* c is annoying */
		.pamh = pamh,
		.sockname = PAM_OKTA_SOCKNAME,
		.sshd = "sshd",
		.mode = OKTA_MODE_MFA_OOB,
	};
	struct state *st = &_st;
	int rv;
	int s;

	pam_syslog(pamh, LOG_DEBUG, "%s", __func__);

	if (parse_args(st, argc, argv) == -1)
		return (PAM_SERVICE_ERR);

	rv = okta_authn_prepare(st);
	if (rv != PAM_SUCCESS)
		return (rv);

	s = okta_connect(st);
	if (s == -1)
		return (PAM_SERVICE_ERR);

	rv = okta_authn(st, s);

	do {
		rv = okta_loop(pamh, st, s);
	} while (rv == PAM_INCOMPLETE);

	close(s);

	return (rv);
}

PAM_EXTERN int
pam_sm_setcred(pam_handle_t *pamh, int flags,
    int argc, const char **argv)
{
	int rv = PAM_SUCCESS;

	pam_syslog(pamh, LOG_DEBUG,
	    "%s: %s", __func__, pam_strerror(pamh, rv));

	return (rv);
}

PAM_EXTERN int
pam_sm_acct_mgmt(pam_handle_t *pamh, int flags,
    int argc, const char *argv[])
{
	int rv = PAM_SUCCESS;

	pam_syslog(pamh, LOG_DEBUG,
	    "%s: %s", __func__, pam_strerror(pamh, rv));

	return (rv);
}

PAM_EXTERN int
pam_sm_open_session(pam_handle_t *pamh, int flags,
    int argc, const char *argv[])
{
	int rv = PAM_SUCCESS;

	pam_syslog(pamh, LOG_DEBUG,
	    "%s: %s", __func__, pam_strerror(pamh, rv));

	return (rv);
}

PAM_EXTERN int
pam_sm_close_session(pam_handle_t *pamh, int flags,
    int argc, const char **argv)
{
	int rv = PAM_SUCCESS;

	pam_syslog(pamh, LOG_DEBUG,
	    "%s: %s", __func__, pam_strerror(pamh, rv));

	return (rv);
}

PAM_EXTERN int
pam_sm_chauthtok(pam_handle_t *pamh, int flags,
    int argc, const char *argv[])
{
	int rv = PAM_SERVICE_ERR;

	pam_syslog(pamh, LOG_DEBUG,
	    "%s: %s", __func__, pam_strerror(pamh, rv));

	return (rv);
}
